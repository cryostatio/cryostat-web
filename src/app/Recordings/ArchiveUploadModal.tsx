/*
 * Copyright The Cryostat Authors
 * 
 * The Universal Permissive License (UPL), Version 1.0
 * 
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 * 
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 * 
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 * 
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { ActionGroup, Button, FileUpload, Form, FormGroup, Modal, ModalVariant } from '@patternfly/react-core';
import { first, tap } from 'rxjs/operators';
import { ServiceContext } from '@app/Shared/Services/Services';
import { NotificationsContext } from '@app/Notifications/Notifications';

export interface ArchiveUploadModalProps {
  visible: boolean;
  onClose: () => void;
}

export const ArchiveUploadModal: React.FunctionComponent<ArchiveUploadModalProps> = props => {
  const context = React.useContext(ServiceContext);
  const notifications = React.useContext(NotificationsContext);
  const [uploadFile, setUploadFile] = React.useState(undefined as File | undefined);
  const [filename, setFilename] = React.useState('' as string | undefined);
  const [uploading, setUploading] = React.useState(false);
  const [rejected, setRejected] = React.useState(false);

  const reset = () => {
    setUploadFile(undefined);
    setFilename('');
    setUploading(false);
    setRejected(true);
  };

  const handleFileChange = (file, filename) => {
    setRejected(false);
    setUploadFile(file);
    setFilename(filename);
  };

  const handleReject = () => {
    setRejected(true);
  };

  const handleClose = () => {
    reset();
    props.onClose();
  };

  const handleSubmit = () => {
    if (!uploadFile) {
      notifications.warning('Attempted to submit JFR upload without a file selected');
      return;
    }
    setUploading(true);
    context.api.uploadRecording(uploadFile)
      .pipe(
        first(),
        tap(() => setUploading(false)),
      )
      .subscribe(handleClose, reset);
  };

  return (
    <Modal
      isOpen={props.visible}
      variant={ModalVariant.large}
      showClose={true}
      onClose={handleClose}
      title="Re-Upload Archived Recording"
      description="Select a JDK Flight Recorder file to re-upload. Files must be .jfr binary format and follow the naming convention used by Cryostat when archiving recordings."
      >
      <Form>
        <FormGroup
          label="JFR File"
          isRequired
          fieldId="file"
          validated={rejected ? 'error' : 'default'}
        >
          <FileUpload
            id="file-upload"
            value={uploadFile}
            filename={filename}
            onChange={handleFileChange}
            isLoading={uploading}
            validated={rejected ? 'error' : 'default'}
            dropzoneProps={{
              accept: '.jfr',
              onDropRejected: handleReject
            }}
          />
        </FormGroup>
        <ActionGroup>
          <Button variant="primary" onClick={handleSubmit} isDisabled={!filename}>Submit</Button>
          <Button variant="link" onClick={handleClose}>Cancel</Button>
        </ActionGroup>
      </Form>
    </Modal>
  );
};
