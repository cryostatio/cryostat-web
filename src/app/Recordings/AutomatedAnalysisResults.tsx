/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CustomRecordingFormData } from '@app/CreateRecording/types';
import {
  ClickableAutomatedAnalysisLabel,
  clickableAutomatedAnalysisKey,
} from '@app/Dashboard/AutomatedAnalysis/ClickableAutomatedAnalysisLabel';
import { CryostatLink } from '@app/Shared/Components/CryostatLink';
import {
  AggregateReport,
  AnalysisResult,
  CategorizedRuleEvaluations,
  NotificationCategory,
  NullableTarget,
  Target,
  TemplateType,
} from '@app/Shared/Services/api.types';
import { ServiceContext } from '@app/Shared/Services/Services';
import useDayjs from '@app/utils/hooks/useDayjs';
import { useSubscriptions } from '@app/utils/hooks/useSubscriptions';
import { useCryostatTranslation } from '@i18n/i18nextUtil';
import {
  Bullseye,
  Grid,
  GridItem,
  LabelGroup,
  Spinner,
  EmptyState,
  EmptyStateHeader,
  EmptyStateIcon,
  Text,
  Stack,
  StackItem,
  Button,
} from '@patternfly/react-core';
import { ProcessAutomationIcon, SearchIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { Trans } from 'react-i18next';
import { Observable } from 'rxjs';
import { concatMap, filter, map, tap } from 'rxjs/operators';

export interface TargetAnalysisProps {
  target: NullableTarget;
  refreshRequest?: Observable<void>;
  immediate?: boolean;
}

export const TargetAnalysis: React.FC<TargetAnalysisProps> = ({ target, refreshRequest, immediate }) => {
  const { t } = useCryostatTranslation();
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions([target]);
  const [loading, setLoading] = React.useState(false);
  const [sourceCount, setSourceCount] = React.useState(0);
  const [report, setReport] = React.useState(undefined as AggregateReport | undefined);

  const emptyReport = React.useMemo(() => !report?.aggregate?.count, [report]);
  const hasSources = React.useMemo(() => !!target && sourceCount > 0, [target, sourceCount]);

  const fetchReport = React.useCallback(
    (target: Target) => {
      setLoading(true);
      return context.api.getCurrentReportForTarget(target).pipe(tap(() => setLoading(false)));
    },
    [setLoading, context.api],
  );

  React.useEffect(() => {
    if (!target) {
      return;
    }
    addSubscription(
      context.api
        .getTargetActiveRecordings(target)
        .pipe(
          map((a) => a.filter((r) => !r.name.toLowerCase().startsWith('snapshot'))),
          map((a) => a.length),
        )
        .subscribe((count) => {
          setSourceCount(count);

          addSubscription(
            context.notificationChannel
              .messages(NotificationCategory.ActiveRecordingCreated)
              .pipe(
                filter((msg) => msg.message.jvmId === target.jvmId),
                filter((msg) => !msg.message.recording.name.toLowerCase().startsWith('snapshot')),
              )
              .subscribe(() => setSourceCount((prev) => prev + 1)),
          );
          addSubscription(
            context.notificationChannel
              .messages(NotificationCategory.ActiveRecordingDeleted)
              .pipe(
                filter((msg) => msg.message.jvmId === target.jvmId),
                filter((msg) => !msg.message.recording.name.toLowerCase().startsWith('snapshot')),
              )
              .subscribe(() => setSourceCount((prev) => prev - 1)),
          );
        }),
    );
  }, [target, addSubscription, context.api, context.notificationChannel, setSourceCount]);

  const handleRefresh = React.useCallback(() => {
    if (!target || !hasSources) {
      return;
    }
    setLoading(true);
    // this will trigger a ReportSuccess notification which we are listening for,
    // and the response body with the job ID is not particularly relevant
    addSubscription(
      context.api
        .sendRequest('v4.1', `/targets/${target.id}/reports`, {
          method: 'POST',
        })
        .subscribe(),
    );
  }, [target, hasSources, addSubscription, context.api, setLoading]);

  React.useEffect(() => {
    if (!!target && immediate && hasSources) {
      handleRefresh();
    }
  }, [target, immediate, hasSources, handleRefresh]);

  React.useEffect(() => {
    if (!target || !refreshRequest) {
      return;
    }
    addSubscription(refreshRequest.subscribe(() => handleRefresh()));
  }, [target, addSubscription, refreshRequest, handleRefresh]);

  React.useEffect(() => {
    if (!target) {
      return;
    }
    addSubscription(fetchReport(target).subscribe((report) => setReport(report)));
  }, [target, addSubscription, fetchReport, setReport]);

  React.useEffect(() => {
    if (!target) {
      return;
    }
    addSubscription(
      context.notificationChannel
        .messages(NotificationCategory.ReportSuccess)
        .pipe(
          filter((msg) => msg.message.jvmId === target.jvmId),
          concatMap(() => fetchReport(target)),
        )
        .subscribe((report) => setReport(report)),
    );
  }, [target, context.notificationChannel, addSubscription, fetchReport]);

  const categorizedEvaluations = React.useMemo(() => {
    if (loading || emptyReport) {
      return [];
    }
    const map = new Map<string, AnalysisResult[]>();
    report!
      .data!.map((e) => e.value)
      .forEach((evaluation) => {
        const topicValue = map.get(evaluation.topic);
        if (topicValue === undefined) {
          map.set(evaluation.topic, [evaluation]);
        } else {
          topicValue.push(evaluation);
          topicValue.sort((a, b) => b.score - a.score || a.name.localeCompare(b.name));
        }
      });
    return (Array.from(map) as CategorizedRuleEvaluations[]).sort();
  }, [loading, emptyReport, report]);

  return (
    <>
      {!target ? (
        <EmptyState>
          <EmptyStateHeader
            titleText={t('AutomatedAnalysisResults.REPORT_UNAVAILABLE')}
            icon={<EmptyStateIcon icon={SearchIcon} />}
            headingLevel="h4"
          />
        </EmptyState>
      ) : !hasSources ? (
        <Text>
          <Trans
            t={t}
            components={[
              <CryostatLink
                state={
                  {
                    name: 'analysis',
                    continuous: true,
                    restart: true,
                    template: { name: 'Continuous', type: 'TARGET' as TemplateType },
                    maxAge: 10,
                    maxAgeUnit: 60,
                    maxSize: 20,
                    maxSizeUnit: 1024 * 1024,
                    toDisk: true,
                  } as Partial<CustomRecordingFormData>
                }
                to="/recordings/create"
              />,
            ]}
          >
            AutomatedAnalysisResults.CREATE_RECORDING
          </Trans>
        </Text>
      ) : undefined}
      {loading ? (
        <Bullseye>
          <Spinner />
        </Bullseye>
      ) : (
        <AutomatedAnalysisResults
          timestamp={report?.lastUpdated}
          analyses={categorizedEvaluations}
          canRefresh={!!target && hasSources}
          handleRefresh={handleRefresh}
        />
      )}
    </>
  );
};

export interface AutomatedAnalysisResultsProps {
  timestamp?: number;
  analyses: CategorizedRuleEvaluations[];
  canRefresh: boolean;
  handleRefresh: () => void;
}

export const AutomatedAnalysisResults: React.FC<AutomatedAnalysisResultsProps> = ({
  timestamp,
  analyses,
  handleRefresh,
  canRefresh,
}) => {
  const { t } = useCryostatTranslation();
  const [dayjs, dateTimeFormat] = useDayjs();
  return (
    <>
      {!analyses.length ? (
        <EmptyState>
          <EmptyStateHeader
            titleText={t('AutomatedAnalysisResults.REPORT_UNAVAILABLE')}
            icon={<EmptyStateIcon icon={SearchIcon} />}
            headingLevel="h4"
          />
          <Button onClick={handleRefresh} isDisabled={!canRefresh}>
            <ProcessAutomationIcon />
          </Button>
        </EmptyState>
      ) : (
        <>
          <Stack hasGutter>
            <StackItem>
              <Text>
                {t('AutomatedAnalysisResults.LAST_UPDATE', {
                  datetime: timestamp
                    ? dayjs(new Date(timestamp * 1000))
                        .tz(dateTimeFormat.timeZone.full)
                        .format('LLLL')
                    : 'unknown',
                })}
              </Text>
            </StackItem>
            <StackItem isFilled>
              <Grid>
                {
                  // TODO port over dashboard card name and score filter controls
                }
                {analyses.map(([topic, evaluations]) => (
                  <GridItem className="automated-analysis-grid-item" span={2} key={`gridItem-${topic}`}>
                    <LabelGroup
                      className="automated-analysis-topic-label-groups"
                      categoryName={topic}
                      isVertical
                      numLabels={2}
                      isCompact
                      key={topic}
                    >
                      {evaluations.map((evaluation) => (
                        <ClickableAutomatedAnalysisLabel result={evaluation} key={clickableAutomatedAnalysisKey} />
                      ))}
                    </LabelGroup>
                  </GridItem>
                ))}
              </Grid>
            </StackItem>
          </Stack>
        </>
      )}
    </>
  );
};
