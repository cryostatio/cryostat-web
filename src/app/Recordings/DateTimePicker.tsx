/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


import {
  Button,
  ButtonVariant,
  DatePicker,
  Flex,
  FlexItem,
  InputGroup,
  isValidDate,
  Text,
  TimePicker,
  yyyyMMddFormat,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import React from 'react';

export interface DateTimePickerProps {
  onSubmit: (startDate) => void;
}

export const DateTimePicker: React.FunctionComponent<DateTimePickerProps> = (props) => {
  const [start, setStart] = React.useState(new Date(0));
  const [searchDisabled, setSearchDisabled] = React.useState(true);

  const onStartDateChange = React.useCallback(
    (inputDate, newStartDate) => {
      if (isValidDate(start) && isValidDate(newStartDate) && inputDate === yyyyMMddFormat(newStartDate)) {
        setStart(new Date(newStartDate));
        setSearchDisabled(false);
      } else {
        setSearchDisabled(true);
      }
    },
    [start, isValidDate, yyyyMMddFormat]
  );

  const onStartTimeChange = React.useCallback(
    (unused, hour, minute) => {
      let updated = new Date(start);
      updated.setUTCHours(hour, minute);
      setStart(updated);
    },
    [start, setStart, isValidDate]
  );

  const handleSubmit = React.useCallback(() => {
    props.onSubmit(`${start.toISOString()}`);
  }, [start, props.onSubmit]);

  return (
    <Flex>
      <FlexItem>
        <InputGroup>
          <DatePicker
            onChange={onStartDateChange}
            aria-label="Start date"
            placeholder="YYYY-MM-DD" />
          <TimePicker is24Hour aria-label="Start time" className="time-picker" onChange={onStartTimeChange} />
        </InputGroup>
        </FlexItem>
        <FlexItem>
          <Text>UTC</Text>
        </FlexItem>
        <FlexItem>
          <Button variant={ButtonVariant.control} aria-label="search button for start date" isDisabled={searchDisabled} onClick={handleSubmit}>
            <SearchIcon />
          </Button>
      </FlexItem>
    </Flex>
  );
};
