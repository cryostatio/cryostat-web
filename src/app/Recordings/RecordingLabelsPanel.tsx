/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BulkEditLabels } from '@app/RecordingMetadata/BulkEditLabels';
import { RecordingDirectory, ArchivedRecording } from '@app/Shared/Services/api.types';
import {
  DrawerActions,
  DrawerCloseButton,
  DrawerHead,
  DrawerPanelBody,
  DrawerPanelContent,
} from '@patternfly/react-core';
import React from 'react';

export interface RecordingLabelsPanelProps {
  setShowPanel: (showPanel: React.SetStateAction<boolean>) => void;
  isTargetRecording: boolean;
  isUploadsTable?: boolean;
  checkedIndices: number[];
  directory?: RecordingDirectory;
  directoryRecordings?: ArchivedRecording[];
}

export const RecordingLabelsPanel: React.FC<RecordingLabelsPanelProps> = (props) => {
  return (
    <DrawerPanelContent isResizable>
      <DrawerHead>
        <DrawerActions>
          <DrawerCloseButton
            onClick={() => props.setShowPanel(false)}
            data-testid="hide-table-actions-panel"
            aria-label="hide table actions panel"
          />
        </DrawerActions>
      </DrawerHead>
      <DrawerPanelBody>
        <BulkEditLabels
          isTargetRecording={props.isTargetRecording}
          checkedIndices={props.checkedIndices}
          isUploadsTable={props.isUploadsTable}
          directory={props.directory}
          directoryRecordings={props.directoryRecordings}
        />
      </DrawerPanelBody>
    </DrawerPanelContent>
  );
};
