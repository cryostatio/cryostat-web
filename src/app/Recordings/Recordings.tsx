/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ServiceContext } from '@app/Shared/Services/Services';
import { TargetView } from '@app/TargetView/TargetView';
import { useSubscriptions } from '@app/utils/useSubscriptions';
import { getActiveTab, switchTab } from '@app/utils/utils';
import { Card, CardBody, CardTitle, Tab, Tabs, TabTitleText } from '@patternfly/react-core';
import * as React from 'react';
import { useHistory, useLocation } from 'react-router';
import { ActiveRecordingsTable } from './ActiveRecordingsTable';
import { ArchivedRecordingsTable } from './ArchivedRecordingsTable';

enum RecordingTab {
  ACTIVE_RECORDING = 'active-recording',
  ARCHIVED_RECORDING = 'archived-recording',
}

export interface RecordingsProps {}

export const Recordings: React.FC<RecordingsProps> = ({ ...props }) => {
  const { search, pathname } = useLocation();
  const history = useHistory();
  const context = React.useContext(ServiceContext);
  const addSubscription = useSubscriptions();

  const activeTab = React.useMemo(() => {
    return getActiveTab(search, 'tab', Object.values(RecordingTab), RecordingTab.ACTIVE_RECORDING);
  }, [search]);

  const [archiveEnabled, setArchiveEnabled] = React.useState(false);

  React.useEffect(() => {
    addSubscription(context.api.isArchiveEnabled().subscribe(setArchiveEnabled));
  }, [context.api, addSubscription, setArchiveEnabled]);

  const onTabSelect = React.useCallback(
    (_: React.MouseEvent, key: string | number) =>
      switchTab(history, pathname, search, { tabKey: 'tab', tabValue: `${key}` }),
    [history, pathname, search],
  );

  const targetAsObs = React.useMemo(() => context.target.target(), [context.target]);

  const cardBody = React.useMemo(() => {
    return archiveEnabled ? (
      <Tabs id="recordings" activeKey={activeTab} onSelect={onTabSelect} unmountOnExit>
        <Tab
          id="active-recordings"
          eventKey={RecordingTab.ACTIVE_RECORDING}
          title={<TabTitleText>Active Recordings</TabTitleText>}
          data-quickstart-id="active-recordings-tab"
        >
          <ActiveRecordingsTable archiveEnabled={true} />
        </Tab>
        <Tab
          id="archived-recordings"
          eventKey={RecordingTab.ARCHIVED_RECORDING}
          title={<TabTitleText>Archived Recordings</TabTitleText>}
          data-quickstart-id="archived-recordings-tab"
        >
          <ArchivedRecordingsTable target={targetAsObs} isUploadsTable={false} isNestedTable={false} />
        </Tab>
      </Tabs>
    ) : (
      <>
        <CardTitle>Active Recordings</CardTitle>
        <ActiveRecordingsTable archiveEnabled={false} />
      </>
    );
  }, [archiveEnabled, activeTab, onTabSelect, targetAsObs]);

  return (
    <TargetView {...props} pageTitle="Recordings">
      <Card>
        <CardBody>{cardBody}</CardBody>
      </Card>
    </TargetView>
  );
};

export default Recordings;
