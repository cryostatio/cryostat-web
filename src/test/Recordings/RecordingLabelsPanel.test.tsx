/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

jest.mock('@app/RecordingMetadata/BulkEditLabels', () => {
  return {
    BulkEditLabels: jest.fn((props) => {
      return (
        <div>
          Bulk Edit Labels
          {props.children}
        </div>
      );
    }),
  };
});

import { RecordingLabelsPanel } from '@app/Recordings/RecordingLabelsPanel';
import { ArchivedRecording } from '@app/Shared/Services/Api.service';
import { Drawer, DrawerContent } from '@patternfly/react-core';
import { renderDefault } from '../Common';

const mockRecordingLabels = {
  someLabel: 'someValue',
};

const mockRecording: ArchivedRecording = {
  name: 'someRecording',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  size: 2048,
  archivedTime: 2048,
};

describe('<RecordingLabelsPanel />', () => {
  let isTargetRecording = true;
  let checkedIndices = [];
  let recordings = [mockRecording];

  const props = {
    setShowPanel: jest.fn(() => (showPanel: boolean) => {}),
    isTargetRecording: isTargetRecording,
    checkedIndices: checkedIndices,
    recordings: recordings,
  };

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <Drawer isExpanded={true} isInline>
          <DrawerContent panelContent={<RecordingLabelsPanel {...props} />} />
        </Drawer>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('displays the bulk labels editor within the resizeable drawer panel', async () => {
    renderDefault(
      <Drawer isExpanded={true} isInline>
        <DrawerContent panelContent={<RecordingLabelsPanel {...props} />} />
      </Drawer>
    );
    expect(screen.getByText('Bulk Edit Labels')).toBeInTheDocument();
    expect(screen.getAllByLabelText('Resize').length).toBe(1);
    expect(screen.getAllByLabelText('hide table actions panel').length).toBe(1);
  });
});
