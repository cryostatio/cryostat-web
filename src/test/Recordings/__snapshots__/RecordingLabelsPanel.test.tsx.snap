// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecordingLabelsPanel /> renders correctly 1`] = `
<div
  className="pf-v5-c-drawer pf-m-expanded pf-m-inline"
>
  <div
    className="pf-v5-c-drawer__main"
  >
    <div
      className="pf-v5-c-drawer__content"
    />
    <div
      className="pf-v5-c-drawer__panel pf-m-resizable"
      hidden={false}
      id="pf-drawer-panel-0"
      onTransitionEnd={[Function]}
    >
      <div
        aria-controls="pf-drawer-panel-0"
        aria-label="Resize"
        aria-orientation="vertical"
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        className="pf-v5-c-drawer__splitter pf-m-vertical"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="separator"
        tabIndex={0}
      >
        <div
          aria-hidden={true}
          className="pf-v5-c-drawer__splitter-handle"
        />
      </div>
      <div
        className="pf-v5-c-drawer__panel-main"
      >
        <div
          className="pf-v5-c-drawer__body"
        >
          <div
            className="pf-v5-c-drawer__head"
          >
            <div
              className="pf-v5-c-drawer__actions"
            >
              <div
                className="pf-v5-c-drawer__close"
                data-testid="hide-table-actions-panel"
                onClick={[Function]}
              >
                <button
                  aria-disabled={false}
                  aria-label="hide table actions panel"
                  className="pf-v5-c-button pf-m-plain"
                  data-ouia-component-id="OUIA-Generated-Button-plain-1"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe={true}
                  disabled={false}
                  onClick={[Function]}
                  role={null}
                  type="button"
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    className="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 352 512"
                    width="1em"
                  >
                    <path
                      d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pf-v5-c-drawer__body"
        >
          <div>
            Bulk Edit Labels
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
